kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerManager.java

```diff
@@ -274,33 +312,99 @@ public String createConsumer(String group, ConsumerInstanceConfig instanceConfig
+                // like the new consumer does. When the new consumer is available we may be
+                // able to update this to provide better feedback to the user.
+                Exception responseException = e;
+                if (!(e instanceof RestException)) {
+                  responseException = Errors.kafkaErrorException(e);
+                }
+                callback.onCompletion(null, responseException);
+              } else {
+                callback.onCompletion(records, null);
               }
-              callback.onCompletion(null, responseException);
-            } else {
-              callback.onCompletion(records, null);
             }
           }
-        }
     );
+    return executor.submit(new RunnableReadTask(new ReadTaskState(task, state, callback)));
+  }
+
+  class RunnableReadTask implements Runnable, Delayed {
+    private final ReadTaskState taskState;
+    private final KafkaRestConfig consumerConfig;
+    private final long started;
+    private final long requestExpiration;
+    // Expiration if this task is waiting, considering both the expiration of the whole task and
+    // a single backoff, if one is in progress
+    private long waitExpirationMs;
+
+    public RunnableReadTask(ReadTaskState taskState) {
+      this.taskState = taskState;
+      this.started = config.getTime().milliseconds();
+      this.consumerConfig = taskState.consumerState.getConfig();
+      this.requestExpiration = this.started
+              + consumerConfig.getInt(KafkaRestConfig.CONSUMER_REQUEST_TIMEOUT_MS_CONFIG);
+      this.waitExpirationMs = 0;
+    }
+
+    @Override
+    public void run() {
+      try {
+        log.trace("Executing consumer read task ({})", taskState.task);
+
+        taskState.task.doPartialRead();
+        KafkaConsumerManager.this.updateExpiration(taskState.consumerState);
+        if (!taskState.task.isDone()) {
+          long backoffTime = config.getTime().milliseconds()
+              + consumerConfig.getInt(KafkaRestConfig.CONSUMER_ITERATOR_BACKOFF_MS_CONFIG);
+          waitExpirationMs = Math.min(backoffTime, requestExpiration);
+
+          // add to delayedReadTasks so the scheduler thread can re-schedule another partial read
+          delayedReadTasks.add(this);
+        } else {
+          log.trace("Finished executing consumer read task ({})", taskState.task);
+        }
+      } catch (Exception e) {
+        log.error("Failed to read records consumer "
+                        + taskState.consumerState.getId().toString(),
+                e);
+        Exception responseException = e;
+        if (!(e instanceof RestException)) {
+          responseException = Errors.kafkaErrorException(e);
+        }
+        taskState.callback.onCompletion(null, responseException);
+      }
+    }
+
+    @Override
+    public long getDelay(TimeUnit unit) {
+      return waitExpirationMs - config.getTime().milliseconds();
```

Comment:
To make use of the unit, we could do something like this:
```java
long delayMs = waitExpirationMs - config.getTime().milliseconds();
return TimeUnit.MILLISECONDS.convert(delayMs, unit);
```

Reply:
I did not know it was that easy. For sure